File: apps/mobile/lib/core/services/upload_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart' show debugPrint, ValueNotifier;
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../network/api_client.dart';

class UploadProgress {
  final double progress; // 0.0 - 1.0
  final bool isUploading;
  final String? error;

  const UploadProgress({
    this.progress = 0,
    this.isUploading = false,
    this.error,
  });
}

class UploadResult {
  final String url;
  final String fileKey;
  final int size;
  final String mimeType;

  const UploadResult({
    required this.url,
    required this.fileKey,
    required this.size,
    required this.mimeType,
  });
}

class UploadService {
  final Dio _dio;

  UploadService(this._dio);

  /// Three-step upload flow: presign → PUT → confirm
  Future<UploadResult> uploadFile({
    required File file,
    required String filename,
    required String mimeType,
    required String category, // 'image' | 'file' | 'voice' | 'avatar'
    ValueNotifier<UploadProgress>? progressNotifier,
  }) async {
    progressNotifier?.value = const UploadProgress(
      isUploading: true,
      progress: 0,
    );

    try {
      // Step 1: Get presigned URL
      final presignResponse = await _dio.get(
        '/api/v1/upload/presign',
        queryParameters: {
          'filename': filename,
          'mimeType': mimeType,
          'category': category,
        },
      );

      final uploadUrl = presignResponse.data['uploadUrl'] as String;
      final fileKey = presignResponse.data['fileKey'] as String;

      progressNotifier?.value = const UploadProgress(
        isUploading: true,
        progress: 0.1,
      );

      // Step 2: PUT file directly to S3/MinIO
      final fileBytes = await file.readAsBytes();
      final uploadDio = Dio(); // Separate Dio instance (no auth headers)
      await uploadDio.put(
        uploadUrl,
        data: Stream.fromIterable([fileBytes]),
        options: Options(
          headers: {
            'Content-Type': mimeType,
            'Content-Length': fileBytes.length,
          },
        ),
        onSendProgress: (sent, total) {
          if (total > 0) {
            final progress = 0.1 + 0.7 * (sent / total);
            progressNotifier?.value = UploadProgress(
              isUploading: true,
              progress: progress,
            );
          }
        },
      );

      progressNotifier?.value = const UploadProgress(
        isUploading: true,
        progress: 0.85,
      );

      // Step 3: Confirm upload
      final confirmResponse = await _dio.post(
        '/api/v1/upload/confirm',
        data: {'fileKey': fileKey},
      );

      final result = UploadResult(
        url: confirmResponse.data['url'] as String,
        fileKey: confirmResponse.data['fileKey'] as String,
        size: confirmResponse.data['size'] as int,
        mimeType: confirmResponse.data['mimeType'] as String,
      );

      progressNotifier?.value = const UploadProgress(
        isUploading: false,
        progress: 1.0,
      );

      return result;
    } catch (e) {
      debugPrint('[UploadService] Upload failed: $e');
      progressNotifier?.value = UploadProgress(
        isUploading: false,
        error: e.toString(),
      );
      rethrow;
    }
  }
}

final uploadServiceProvider = Provider<UploadService>((ref) {
  final dio = ref.read(dioProvider);
  return UploadService(dio);
});

File: apps/desktop/src/renderer/services/uploadService.ts
/**
 * Upload service: presign → PUT → confirm flow
 */
import { API_BASE_URL } from '@renderer/config';

export interface UploadResult {
  url: string;
  fileKey: string;
  size: number;
  mimeType: string;
}

export interface UploadProgressCallback {
  (progress: number): void; // 0.0 - 1.0
}

const SIZE_LIMITS: Record<string, number> = {
  image: 10 * 1024 * 1024, // 10MB
  file: 50 * 1024 * 1024,  // 50MB
  avatar: 5 * 1024 * 1024, // 5MB
  voice: 10 * 1024 * 1024, // 10MB
};

export async function uploadFile(
  file: File,
  category: 'image' | 'file' | 'avatar' | 'voice',
  onProgress?: UploadProgressCallback,
): Promise<UploadResult> {
  const token = await window.electronAPI.getToken();
  if (!token) throw new Error('Not authenticated');

  // Check size limit
  const limit = SIZE_LIMITS[category] ?? SIZE_LIMITS.file;
  if (file.size > limit) {
    throw new Error(`File too large (max ${Math.round(limit / 1024 / 1024)}MB)`);
  }

  const mimeType = file.type || 'application/octet-stream';

  onProgress?.(0.05);

  // Step 1: Presign
  const presignRes = await fetch(
    `${API_BASE_URL}/api/v1/upload/presign?filename=${encodeURIComponent(file.name)}&mimeType=${encodeURIComponent(mimeType)}&category=${category}`,
    { headers: { Authorization: `Bearer ${token}` } },
  );
  if (!presignRes.ok) {
    const err = await presignRes.text();
    throw new Error(`Presign failed: ${err}`);
  }
  const { uploadUrl, fileKey } = await presignRes.json();

  onProgress?.(0.1);

  // Step 2: PUT to S3/MinIO
  const xhr = new XMLHttpRequest();
  await new Promise<void>((resolve, reject) => {
    xhr.open('PUT', uploadUrl);
    xhr.setRequestHeader('Content-Type', mimeType);

    xhr.upload.onprogress = (e) => {
      if (e.lengthComputable) {
        onProgress?.(0.1 + 0.7 * (e.loaded / e.total));
      }
    };
    xhr.onload = () => {
      if (xhr.status >= 200 && xhr.status < 300) {
        resolve();
      } else {
        reject(new Error(`Upload failed: ${xhr.status}`));
      }
    };
    xhr.onerror = () => reject(new Error('Upload network error'));
    xhr.send(file);
  });

  onProgress?.(0.85);

  // Step 3: Confirm
  const confirmRes = await fetch(
    `${API_BASE_URL}/api/v1/upload/confirm`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ fileKey }),
    },
  );
  if (!confirmRes.ok) {
    const err = await confirmRes.text();
    throw new Error(`Confirm failed: ${err}`);
  }
  const result = await confirmRes.json();

  onProgress?.(1.0);

  return {
    url: result.url,
    fileKey: result.fileKey,
    size: result.size,
    mimeType: result.mimeType,
  };
}
